magicblock-chainlink/src/chainlink/fetch_cloner.rs

```diff
@@ -0,0 +1,2609 @@
+    pub missing_delegation_record: Vec<(Pubkey, u64)>,
+}
+
+impl FetchAndCloneResult {
+    pub fn pubkeys_not_found_on_chain(&self) -> Vec<Pubkey> {
+        self.not_found_on_chain.iter().map(|(p, _)| *p).collect()
+    }
+
+    pub fn pubkeys_missing_delegation_record(&self) -> Vec<Pubkey> {
+        self.missing_delegation_record
+            .iter()
+            .map(|(p, _)| *p)
+            .collect()
+    }
+
+    pub fn is_ok(&self) -> bool {
+        self.not_found_on_chain.is_empty()
+            && self.missing_delegation_record.is_empty()
+    }
+}
+
+impl fmt::Display for FetchAndCloneResult {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.is_ok() {
+            write!(f, "All accounts fetched and cloned successfully")
+        } else {
+            if !self.not_found_on_chain.is_empty() {
+                writeln!(
+                    f,
+                    "Accounts not found on chain: {:?}",
+                    self.not_found_on_chain
+                        .iter()
+                        .map(|(p, _)| p.to_string())
+                        .collect::<Vec<_>>()
+                )?;
+            }
+            if !self.missing_delegation_record.is_empty() {
+                writeln!(
+                    f,
+                    "Accounts missing delegation record: {:?}",
+                    self.missing_delegation_record
+                        .iter()
+                        .map(|(p, _)| p.to_string())
+                        .collect::<Vec<_>>()
+                )?;
+            }
+            Ok(())
+        }
+    }
+}
+
+impl<T, U, V, C> FetchCloner<T, U, V, C>
+where
+    T: ChainRpcClient,
+    U: ChainPubsubClient,
+    V: AccountsBank,
+    C: Cloner,
+{
+    /// Create FetchCloner with subscription updates properly connected
+    pub fn new(
+        remote_account_provider: &Arc<RemoteAccountProvider<T, U>>,
+        accounts_bank: &Arc<V>,
+        cloner: &Arc<C>,
+        validator_pubkey: Pubkey,
+        faucet_pubkey: Pubkey,
+        subscription_updates_rx: mpsc::Receiver<ForwardedSubscriptionUpdate>,
+    ) -> Self {
+        let blacklisted_accounts =
+            blacklisted_accounts(&validator_pubkey, &faucet_pubkey);
+        let me = Self {
+            remote_account_provider: remote_account_provider.clone(),
+            accounts_bank: accounts_bank.clone(),
+            cloner: cloner.clone(),
+            validator_pubkey,
+            pending_requests: Arc::new(Mutex::new(HashMap::new())),
+            fetch_count: Arc::new(AtomicU64::new(0)),
+            blacklisted_accounts,
+        };
+
+        me.start_subscription_listener(subscription_updates_rx);
+
+        me
+    }
+
+    /// Get the current fetch count
+    pub fn fetch_count(&self) -> u64 {
+        self.fetch_count.load(Ordering::Relaxed)
+    }
+
+    /// Start listening to subscription updates
+    pub fn start_subscription_listener(
+        &self,
```

Comment:
nit: instead of doing bunch of field clones, it's better to use self: Arc<Self>, pattern here, and move the self into async task. This allows to avoid Self::method pattern with bunch fields comming from self anyway.